src/webgpu/shader/validation/statement/phony.spec.ts

```diff
@@ -0,0 +1,131 @@
+export const description = `Validation for phony assignment statements`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { keysOf } from '../../../../common/util/data_tables.js';
+import { scalarTypeOf, Type } from '../../../util/conversion.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+interface Case {
+  value: string;
+  pass: boolean;
+  usesF16?: boolean;
+  gdecl?: string;
+  ldecl?: string;
+}
+
+const kConstructibleTypes = [
+  'bool',
+  'i32',
+  'u32',
+  'f32',
+  'f16',
+  'vec2f',
+  'vec3h',
+  'vec4u',
+  'vec3b',
+  'mat2x3f',
+  'mat4x2h',
+  'abstractInt',
+  'abstractFloat',
+] as const;
+
+const kConstructibleCases = {
+  ...kConstructibleTypes.reduce(
+    (acc, t) => ({
+      ...acc,
+      [t]: {
+        value: Type[t].create(1).wgsl(),
+        pass: true,
+        usesF16: scalarTypeOf(Type[t]).kind === 'f16',
+      },
+    }),
+    {}
+  ),
+  array: { value: 'array(1,2,3)', pass: true },
+  struct: { value: 'S(1,2)', pass: true, gdecl: 'struct S{ a:u32, b:u32}' },
+  atomic_u32: { value: 'xu', pass: false, gdecl: 'var<workgroup> xu: atomic<u32>;' },
+  atomic_i32: { value: 'xi', pass: false, gdecl: 'var<workgroup> xi: atomic<i32>;' },
+} as const;
+
+g.test('rhs_constructible')
+  .desc(`Test that the rhs of 'phony assignment' can be a constructible.`)
+  .params(u => u.combine('type', keysOf(kConstructibleCases)))
+  .beforeAllSubcases(t => {
+    const c = kConstructibleCases[t.params.type] as Case;
+    if (c.usesF16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const { value, pass, usesF16, gdecl } = kConstructibleCases[t.params.type] as Case;
+    const code = `
+${usesF16 ? 'enable f16;' : ''}
+${gdecl ?? ''}
+fn f() {
+  _ = ${value};
+}`;
+    t.expectCompileResult(pass, code);
+  });
+
+const kVarCases = {
+  storage: { value: 'x', gdecl: '@group(0) @binding(0) var<storage> x: array<u32,1>;', pass: true },
+  storage_unsized: {
+    value: 'x',
+    gdecl: '@group(0) @binding(0) var<storage> x: array<u32>;',
+    pass: false,
+  },
+  storage_atomic: {
+    value: 'x',
+    gdecl: '@group(0) @binding(0) var<storage> x: atomic<u32>;',
```

Comment:
Nit: add `read_write` access mode as required by atomics, so that this doesn't fail for the wrong reason.

Reply:
Done